src/emqx_cm.erl

```diff
@@ -286,44 +295,63 @@ discard_session(ClientId) when is_binary(ClientId) ->
+    -> ok                        %% returned by kick, discard
+     | emqx_session:session()    %% returned by {takeover, 'begin'}
+     | list(emqx_type:deliver()) %% returned by {takeover, 'end'}
+  when Action :: kick | discard | {takeover, 'begin'} | {takeover, 'end'}.
+call_or_kill(Action, ConnMod, Pid) ->
     try
+        Timeout = case Action of
+                      {takeover, _} -> ?T_TAKEOVER;
+                      _ -> ?T_KICK
+                  end,
         %% this is essentailly a gen_server:call implemented in emqx_connection
         %% and emqx_ws_connection.
         %% the handle_call is implemented in emqx_channel
-        ok = apply(ConnMod, call, [Pid, Action, ?T_KICK])
+        apply(ConnMod, call, [Pid, Action, Timeout])
     catch
         _ : noproc -> % emqx_ws_connection: call
-            ok = ?tp(debug, "session_already_gone", #{pid => Pid, action => Action});
+            ok = ?tp(debug, "session_already_gone", #{pid => Pid, action => Action}),
+            return_error_if_action_is_takeover(Action, noproc);
         _ : {noproc, _} -> % emqx_connection: gen_server:call
-            ok = ?tp(debug, "session_already_gone", #{pid => Pid, action => Action});
-        _ : {shutdown, _} ->
-            ok = ?tp(debug, "session_already_shutdown", #{pid => Pid, action => Action});
-        _ : {{shutdown, _}, _} ->
-            ok = ?tp(debug, "session_already_shutdown", #{pid => Pid, action => Action});
+            ok = ?tp(debug, "session_already_gone", #{pid => Pid, action => Action}),
+            return_error_if_action_is_takeover(Action, noproc);
+        _ : Reason = {shutdown, _} ->
+            ok = ?tp(debug, "session_already_shutdown", #{pid => Pid, action => Action}),
+            return_error_if_action_is_takeover(Action, Reason);
+        _ : Reason = {{shutdown, _}, _} ->
+            ok = ?tp(debug, "session_already_shutdown", #{pid => Pid, action => Action}),
+            return_error_if_action_is_takeover(Action, Reason);
         _ : {timeout, {gen_server, call, _}} ->
-            ?tp(warning, "session_kick_timeout",
+            ?tp(warning, "call_session_timeout",
                 #{pid => Pid,
                   action => Action,
                   stale_channel => stale_channel_info(Pid)
                  }),
-            ok = force_kill(Pid);
+            ok = force_kill(Pid),
+            return_error_if_action_is_takeover(Action, timeout);
         _ : Error : St ->
-            ?tp(error, "session_kick_exception",
+            ?tp(error, "call_session_exception",
                 #{pid => Pid,
                   action => Action,
                   reason => Error,
                   stacktrace => St,
                   stale_channel => stale_channel_info(Pid)
                  }),
-            ok = force_kill(Pid)
+            ok = force_kill(Pid),
+            return_error_if_action_is_takeover(Action, Error)
     end.
 
+return_error_if_action_is_takeover({takeover, _}, Reason) ->
```

Comment:
Maybe a batter func name. @zmstone Any suggestions?

Reply:
maybe no need for this function if refactor as suggested above